File: src/JSInterop/Leeax.Web.Components.DOM/ElementService.cs
using Leeax.Web.Components.Abstractions;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Leeax.Web.Components.DOM
{
    public class ElementService : IElementService
    {
        public const string ModuleKey = "__" + nameof(ElementService);
        public const string ModulePath = "./_content/Leeax.Web.Components.DOM/ElementService.min.js";

        private long _actionOutOfElementCallbackId;
        private readonly Dictionary<long, Action> _actionOutOfElementCallbacks = new Dictionary<long, Action>();

        private readonly IJSInProcessObjectReference? _jsInProcessObjectRef;
        private readonly IJSObjectReferenceStore _jsRefStore;

        public ElementService(IJSObjectReferenceStore jsRefStore)
        {
            _jsRefStore = jsRefStore;
            jsRefStore.TryGet(ModuleKey, out _jsInProcessObjectRef);
        }

        public bool InsertMarkup(string parentSelector, string? value, InsertionPosition position, MarkupType type)
        {
            _ = parentSelector ?? throw new ArgumentNullException(nameof(parentSelector));

            return _jsInProcessObjectRef!.Invoke<bool>(
                "insertMarkup",
                parentSelector,
                value,
                MapInsertionPosition(position),
                MapMarkupType(type));
        }

        public bool InsertMarkup(ElementReference parentElement, string? value, InsertionPosition position, MarkupType type)
        {
            return _jsInProcessObjectRef!.Invoke<bool>(
                "insertMarkup",
                parentElement,
                value,
                MapInsertionPosition(position),
                MapMarkupType(type));
        }

        public async ValueTask<bool> InsertMarkupAsync(string parentSelector, string? value, InsertionPosition position, MarkupType type)
        {
            _ = parentSelector ?? throw new ArgumentNullException(nameof(parentSelector));

            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            return await module.InvokeAsync<bool>(
                "insertMarkup",
                parentSelector,
                value,
                MapInsertionPosition(position),
                MapMarkupType(type));
        }

        public async ValueTask<bool> InsertMarkupAsync(ElementReference parentElement, string? value, InsertionPosition position, MarkupType type)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            return await module.InvokeAsync<bool>(
                "insertMarkup",
                parentElement,
                value,
                MapInsertionPosition(position),
                MapMarkupType(type));
        }

        public bool RemoveElement(string selector)
        {
            _ = selector ?? throw new ArgumentNullException(nameof(selector));

            return _jsInProcessObjectRef!.Invoke<bool>(
                "removeElement",
                selector);
        }

        public bool RemoveElement(ElementReference element)
        {
            return _jsInProcessObjectRef!.Invoke<bool>(
                "removeElement",
                element);
        }

        public async ValueTask<bool> RemoveElementAsync(string selector)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            return await module.InvokeAsync<bool>(
                "removeElement",
                selector);
        }

        public async ValueTask<bool> RemoveElementAsync(ElementReference element)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            return await module.InvokeAsync<bool>(
                "removeElement",
                element);
        }

        public void ScrollIntoView(ElementReference element, ScrollIntoViewAlignment block, ScrollIntoViewAlignment inline, bool smooth = false)
        {
            _jsInProcessObjectRef!.InvokeVoid(
                "scrollIntoView",
                element,
                MapScrollIntoViewAlignment(block),
                MapScrollIntoViewAlignment(inline),
                smooth);
        }

        public void ScrollIntoView(ElementReference element, bool smooth = false)
        {
            ScrollIntoView(
                element,
                ScrollIntoViewAlignment.Start,
                ScrollIntoViewAlignment.Nearest,
                smooth);
        }

        public async ValueTask ScrollIntoViewAsync(ElementReference element, ScrollIntoViewAlignment block, ScrollIntoViewAlignment inline, bool smooth = false)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            await module.InvokeVoidAsync(
                "scrollIntoView",
                element,
                MapScrollIntoViewAlignment(block),
                MapScrollIntoViewAlignment(inline),
                smooth);
        }

        public ValueTask ScrollIntoViewAsync(ElementReference element, bool smooth = false)
        {
            return ScrollIntoViewAsync(
                element, 
                ScrollIntoViewAlignment.Start,
                ScrollIntoViewAlignment.Nearest,
                smooth);
        }

        public ElementBoundingClientRect? GetBoundingClientRect(ElementReference element)
        {
            return _jsInProcessObjectRef!.Invoke<ElementBoundingClientRect>(
                "getBoundingClientRect",
                element);
        }

        public async ValueTask<ElementBoundingClientRect?> GetBoundingClientRectAsync(ElementReference element)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            return await module.InvokeAsync<ElementBoundingClientRect>(
                "getBoundingClientRect",
                element);
        }

        public ElementPositionInfo? GetPosition(ElementReference element)
        {
            return _jsInProcessObjectRef!.Invoke<ElementPositionInfo>(
                "getPosition",
                element);
        }

        public async ValueTask<ElementPositionInfo?> GetPositionAsync(ElementReference element)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            return await module.InvokeAsync<ElementPositionInfo>(
                "getPosition",
                element);
        }

        private static string MapScrollIntoViewAlignment(ScrollIntoViewAlignment value)
        {
            return value switch
            {
                ScrollIntoViewAlignment.Start => "start",
                ScrollIntoViewAlignment.Center => "center",
                ScrollIntoViewAlignment.End => "end",
                _ => "nearest"
            };
        }

        private static string MapInsertionPosition(InsertionPosition value)
        {
            return value switch
            {
                InsertionPosition.BeforeBegin => "beforebegin",
                InsertionPosition.AfterBegin => "afterbegin",
                InsertionPosition.BeforeEnd => "beforeend",
                InsertionPosition.AfterEnd => "afterend",
                _ => throw new ArgumentException($"Unknown {nameof(InsertionPosition)} \"{nameof(InsertionPosition)}.{value}\".")
            };
        }

        private static string MapMarkupType(MarkupType value)
        {
            return value switch
            {
                MarkupType.HTML => "html",
                MarkupType.Text => "text",
                _ => throw new ArgumentException($"Unknown {nameof(MarkupType)} \"{nameof(MarkupType)}.{value}\".")
            };
        }

        public long AddClickOutsideOfElementHandler(ElementReference[] elements, Action handler)
        {
            _ = handler ?? throw new ArgumentNullException(nameof(handler));

            var id = _actionOutOfElementCallbackId++;

            _actionOutOfElementCallbacks.Add(id, handler);

            _jsInProcessObjectRef!.InvokeVoid(
                "addClickOutsideOfElementHandler",
                DotNetObjectReference.Create(this),
                elements,
                id);

            return id;
        }

        public bool RemoveClickOutsideOfElementHandler(long handlerId)
        {
            if (!_actionOutOfElementCallbacks.ContainsKey(handlerId))
            {
                return false;
            }

            return _jsInProcessObjectRef!.Invoke<bool>(
                "removeClickOutsideOfElementHandler",
                handlerId);
        }

        public async ValueTask<long> AddClickOutsideOfElementHandlerAsync(ElementReference[] elements, Action handler)
        {
            _ = handler ?? throw new ArgumentNullException(nameof(handler));

            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            var id = _actionOutOfElementCallbackId++;

            _actionOutOfElementCallbacks.Add(id, handler);

            await module.InvokeVoidAsync(
                "addClickOutsideOfElementHandler",
                DotNetObjectReference.Create(this),
                elements,
                id);

            return id;
        }

        public async ValueTask<bool> RemoveClickOutsideOfElementHandlerAsync(long handlerId)
        {
            var module = await _jsRefStore
                .ImportOrGetModuleAsync(ModuleKey, ModulePath);

            if (!_actionOutOfElementCallbacks.ContainsKey(handlerId))
            {
                return false;
            }

            return await module.InvokeAsync<bool>(
                "removeClickOutsideOfElementHandler",
                handlerId);
        }

        [JSInvokable]
        public void HandleClickOutsideOfElementCallback(long handlerId)
        {
            if (_actionOutOfElementCallbacks.TryGetValue(handlerId, out var callback))
            {
                _actionOutOfElementCallbacks.Remove(handlerId);

                callback?.Invoke();
                return;
            }

            Console.WriteLine($"[WARN] Unkown 'ClickOutsideOfElement' handler-id '{handlerId}' were passed from javascript.");
        }
    }
}

File: src/JSInterop/Leeax.Web.Components.DOM/TypeScript/ElementService.ts

const _clickOutsideOfElementHandlers: { [id: number]: (e: MouseEvent) => void } = { };

export function insertMarkup(parent: string, value: string, position: InsertPosition, markupType: string): boolean;
export function insertMarkup(parent: HTMLElement, value: string, position: InsertPosition, markupType: string): boolean;
export function insertMarkup(parent: HTMLElement | string, value: string, position: InsertPosition, markupType: string): boolean {

    let target: any = parent;

    if (typeof parent === "string") {
        target = document.querySelector(parent);
    }

    if (!target) {
        return false;
    }

    if (markupType === "html") {
        (<HTMLElement>target).insertAdjacentHTML(position, value);
    } else if (markupType === "text") {
        (<HTMLElement>target).insertAdjacentText(position, value);
    } else {
        return false;
    }

    return true;
}

export function removeElement(element: string): boolean;
export function removeElement(element: HTMLElement): boolean;
export function removeElement(element: HTMLElement | string): boolean {

    let target: any = element;

    if (typeof element === "string") {
        target = document.querySelector(element);
    }

    if (!target) {
        return false;
    }

    // Remove the determined target from its parent
    (<HTMLElement>target).parentElement.removeChild(target);

    return true;
}

//export function isInViewport(element: HTMLElement): boolean {

//    if (!(element instanceof HTMLElement)) {
//        return false;
//    }

//    const rect = element.getBoundingClientRect();
//    const html = document.documentElement;

//    return (
//        rect.top >= 0 &&
//        rect.left >= 0 &&
//        rect.bottom <= (window.innerHeight || html.clientHeight) &&
//        rect.right <= (window.innerWidth || html.clientWidth)
//    );
//}

export function scrollIntoView(element: HTMLElement, block: ScrollLogicalPosition, inline: ScrollLogicalPosition, smooth: boolean): void {

    const options: ScrollIntoViewOptions = {};

    if (block !== null) {
        options.block = block;
    }
    if (inline !== null) {
        options.inline = inline;
    }
    if (smooth) {
        options.behavior = "smooth";
    }

    element.scrollIntoView(options);
}

export function getBoundingClientRect(element: HTMLElement): DOMRect {
    return element.getBoundingClientRect();
}

export function getPosition(element: HTMLElement) {

    const boundingRect = element.getBoundingClientRect();
    const isHidden = boundingRect.left === 0
        && boundingRect.top === 0
        && boundingRect.right === 0
        && boundingRect.bottom === 0;

    return {
        Left: boundingRect.left,
        Top: boundingRect.top,
        Right: isHidden ? 0 : (window.innerWidth - boundingRect.right),
        Bottom: isHidden ? 0 : (window.innerHeight - boundingRect.bottom),
        Height: boundingRect.height,
        Width: boundingRect.width
    };
}

export function addClickOutsideOfElementHandler(callbackTarget: any, elements: HTMLElement[], handlerId: number): boolean {

    if (!elements || !elements.length) return false;

    if (handlerId in _clickOutsideOfElementHandlers) {
        removeClickOutsideOfElementHandler(handlerId);
    }

    const callback = (e: MouseEvent) => {
        if (e.target) {

            for (var i = 0; i < elements.length; i++) {
                if (elements[i].contains(e.target as any)) {
                    return;
                }
            }

            callbackTarget.invokeMethod("HandleClickOutsideOfElementCallback", handlerId);
            removeClickOutsideOfElementHandler(handlerId);
        }
    }

    _clickOutsideOfElementHandlers[handlerId] = callback;

    window.addEventListener("click", callback);

    return true;
}

export function removeClickOutsideOfElementHandler(handlerId: number): boolean {

    if (handlerId in _clickOutsideOfElementHandlers) {
        window.removeEventListener("click", _clickOutsideOfElementHandlers[handlerId]);
        delete _clickOutsideOfElementHandlers[handlerId];
        return true;
    }

    return false;
}